pkg/reconciler/eventtype/eventtype.go

```diff
@@ -0,0 +1,208 @@
+	"fmt"
+	"regexp"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/validation"
+
+	eventingv1alpha1 "github.com/knative/eventing/pkg/apis/eventing/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// Only allow alphanumeric, '-' or '.'.
+var validChars = regexp.MustCompile(`[^-\.a-z0-9]+`)
+
+// Reconciler reconciles EventTypes of different sources.
+type Reconciler struct {
+	Client client.Client
+	Scheme *runtime.Scheme
+}
+
+// ReconcilerArgs are the arguments needed to reconcile EventTypes.
+type ReconcilerArgs struct {
+	EventTypes []*EventTypeArgs
+	Namespace  string
+	Labels     map[string]string
+}
+
+// EventTypeArgs are the arguments needed to create an EventType.
+type EventTypeArgs struct {
+	Type        string
+	Source      string
+	Schema      string
+	Broker      string
+	Description string
+}
+
+// syncEventTypes is a helper struct used to sync EventTypes during the reconciliation process.
+type syncEventTypes struct {
+	ToCreate []eventingv1alpha1.EventType
+	ToDelete []eventingv1alpha1.EventType
+}
+
+// ReconcileEventTypes reconciles the EventTypes taken from 'args', and sets 'owner' as the controller.
+func (r *Reconciler) ReconcileEventTypes(ctx context.Context, owner metav1.Object, args *ReconcilerArgs) error {
+	current, err := r.getEventTypes(ctx, args.Namespace, args.Labels, owner)
+	if err != nil {
+		return err
+	}
+
+	expected, err := r.makeEventTypes(args, owner)
+	if err != nil {
+		return err
+	}
+	sync := r.getEventTypesToSync(current, expected)
+	for _, eventType := range sync.ToCreate {
+		err = r.Client.Create(ctx, &eventType)
+		if err != nil {
+			return err
+		}
+	}
+	// Deletion might not be needed, but just to make sure we dedupe EventTypes
+	// in case some source controller has a bug.
+	for _, eventType := range sync.ToDelete {
+		err = r.Client.Delete(ctx, &eventType)
+		if err != nil && !errors.IsNotFound(err) {
+			return err
+		}
+	}
+	return nil
+}
+
+// getEventTypes returns the EventTypes controlled by 'owner'.
+func (r *Reconciler) getEventTypes(ctx context.Context, namespace string, lbs map[string]string, owner metav1.Object) ([]eventingv1alpha1.EventType, error) {
+	eventTypes := make([]eventingv1alpha1.EventType, 0)
+
+	opts := &client.ListOptions{
+		Namespace:     namespace,
+		LabelSelector: labels.SelectorFromSet(lbs),
+		// Set Raw because if we need to get more than one page, then we will put the continue token
+		// into opts.Raw.Continue.
+		Raw: &metav1.ListOptions{},
+	}
+	for {
+		el := &eventingv1alpha1.EventTypeList{}
+		if err := r.Client.List(ctx, opts, el); err != nil {
+			return nil, err
+		}
+
+		for _, e := range el.Items {
+			if metav1.IsControlledBy(&e, owner) {
+				eventTypes = append(eventTypes, e)
+			}
+		}
+		if el.Continue != "" {
```

Comment:
I think we established elsewhere that if we don't set a `limit` option we don't have to do pagination, so I think this pagination code can be removed. cc @Harwayne

Reply:
Removed, wasn't aware of that. Just copy/pasting from other places

Reply:
:+1: Please also remove these 3 lines starting at https://github.com/knative/eventing-sources/pull/300/commits/dca6574b0ee80007e49633d0756a458c2ca0b9bf#diff-664142aa17d510b591fb692036e3bf87R97:

```go
// Set Raw because if we need to get more than one page, then we will put the continue token
// into opts.Raw.Continue.
Raw: &metav1.ListOptions{},
```

Reply:
Done, thanks!